internal/webhook/managedcluster_webhook.go

```diff
@@ -54,127 +56,264 @@ var (
-	if err := v.Get(ctx, templateRef, template); err != nil {
-		return nil, err
-	}
-	return template, nil
+func (v *ManagedClusterValidator) getManagedClusterTemplate(ctx context.Context, templateNamespace, templateName string) (tpl *hmcv1alpha1.ClusterTemplate, err error) {
+	tpl = new(hmcv1alpha1.ClusterTemplate)
+	return tpl, v.Get(ctx, client.ObjectKey{Namespace: templateNamespace, Name: templateName}, tpl)
 }
 
-func (v *ManagedClusterValidator) isTemplateValid(ctx context.Context, template *v1alpha1.ClusterTemplate) error {
+func (v *ManagedClusterValidator) isTemplateValid(ctx context.Context, template *hmcv1alpha1.ClusterTemplate) error {
 	if !template.Status.Valid {
 		return fmt.Errorf("the template is not valid: %s", template.Status.ValidationError)
 	}
-	err := v.verifyProviders(ctx, template)
-	if err != nil {
-		return fmt.Errorf("providers verification failed: %v", err)
+
+	if err := v.verifyProviders(ctx, template); err != nil {
+		return fmt.Errorf("failed to verify providers: %v", err)
 	}
+
 	return nil
 }
 
-func (v *ManagedClusterValidator) verifyProviders(ctx context.Context, template *v1alpha1.ClusterTemplate) error {
-	requiredProviders := template.Status.Providers
-	management := &v1alpha1.Management{}
-	managementRef := client.ObjectKey{Name: v1alpha1.ManagementName}
-	if err := v.Get(ctx, managementRef, management); err != nil {
+func (v *ManagedClusterValidator) verifyProviders(ctx context.Context, template *hmcv1alpha1.ClusterTemplate) error {
+	management := new(hmcv1alpha1.Management)
+	if err := v.Get(ctx, client.ObjectKey{Name: hmcv1alpha1.ManagementName}, management); err != nil {
 		return err
 	}
 
-	exposedProviders := management.Status.AvailableProviders
-	missingProviders := make(map[string][]string)
-	missingProviders["bootstrap"] = getMissingProviders(exposedProviders.BootstrapProviders, requiredProviders.BootstrapProviders)
-	missingProviders["control plane"] = getMissingProviders(exposedProviders.ControlPlaneProviders, requiredProviders.ControlPlaneProviders)
-	missingProviders["infrastructure"] = getMissingProviders(exposedProviders.InfrastructureProviders, requiredProviders.InfrastructureProviders)
+	const (
+		bootstrapProviderType    = "bootstrap"
+		controlPlateProviderType = "control plane"
+		infraProviderType        = "infrastructure"
+	)
 
-	var errs []error
-	for providerType, missing := range missingProviders {
-		if len(missing) > 0 {
-			sort.Slice(missing, func(i, j int) bool {
-				return missing[i] < missing[j]
-			})
-			errs = append(errs, fmt.Errorf("one or more required %s providers are not deployed yet: %v", providerType, missing))
+	var (
+		exposedProviders  = management.Status.AvailableProviders
+		requiredProviders = template.Status.Providers
+
+		missingBootstrap, missingCP, missingInfra []string
+		wrongVersionProviders                     map[string][]string
+	)
+
+	// on update we have to validate versions between exact the provider tpl and constraints from the cluster tpl
```

Comment:
I've made the check to be performed independently on the request type (effectively on CREATE/UPDATE) but I am not sure that the validation should be implemented in the template-ctrl since the check requires exposed providers which are only available after management object reconciliation, making the controller dependant on another controller which is not suitable IMO

Reply:
I think it still makes sense. Templates should not be valid if they can't be used within the system and it's good to know about it before you assign them. The check itself is quite heavy  for admission and moving it to the controller would also allow us to avoid duplication - we need the same verification when we try to distribute templates through the templatemanagement object and when we validate release objects in the release controller

Reply:
Will do in the follow-up PR